### PR TITLE
style: Split the invalidation processing from the invalidator step.

### PR DESCRIPTION
**Pull request by emilio**
This is the first step in reusing the invalidation machinery for other stuff,
potentially including QuerySelector / QuerySelectorAll.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/18847)

### CODE REVIEW

**Comment by emilio** [comment]
r? @heycam or @SimonSapin 

This is just moving code around, no behavior change (apart from introducing the `InvalidatorCollector` trait).

**Comment by emilio** [comment]
I ended up working more on top of this, so please review each commit individually.

Thanks a lot!

File: components/style/invalidation/element/invalidator.rs [hunk 1]
```diff
@@ -384,23 +377,23 @@ where
...
 
+        let invalidated_descendants = child_invalidator.invalidate_descendants(
+            &invalidations_for_descendants
+        );
+
         // The child may not be a flattened tree child of the current element,
         // but may be arbitrarily deep.
         //
         // Since we keep the traversal flags in terms of the flattened tree,
         // we need to propagate it as appropriate.
```
Comment (heycam) [inline components/style/invalidation/element/invalidator.rs]: This comment seems to be describing what we do inside the `invalidated_descendants` implementation, which could be anything on some abritrary InvalidationProcessor.  This comment is already duplicated in the (one) `invalidated_descendants` implementation so I think we should just drop this comment here.

File: components/style/invalidation/element/invalidator.rs [hunk 2]
```diff
@@ -22,6 +22,11 @@ pub trait InvalidationProcessor<E>
 where
     E: TElement,
 {
+    /// Whether an invalidation that only leaves an eager pseudo-element like
```
Comment (heycam) [inline components/style/invalidation/element/invalidator.rs]: (Not sure what you mean by "leaves".)

**Review by heycam — APPROVED** [review]
Looks good!

**Pull request merged**